### PR TITLE
Fixes function calls / memory access with wasm-implemented host functions 

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

**Comment by mathetake** [comment]
this should makes the #621 to pass the cat_test.go with compilers

File: internal/engine/compiler/impl_amd64.go [hunk 1]
```diff
@@ -5053,15 +5047,19 @@ func (c *amd64Compiler) compileModuleContextInitialization() error {
 	// Note: if there's memory instruction in the function, memory instance must be non-nil.
 	// That is ensured by function validation at module instantiation phase, and that's
 	// why it is ok to skip the initialization if the module's memory instance is nil.
-	if c.ir.HasMemory {
+	if c.ir.HasMemory &&
+		// If this is the host function, our semantic is that host functions use the caller's memory.
```
Comment (codefromthecrypt) [inline internal/engine/compiler/impl_amd64.go:5051]: is this comment relevant anymore it seems not

File: internal/engine/compiler/impl_arm64.go [hunk 2]
```diff
@@ -4198,7 +4192,11 @@ func (c *arm64Compiler) compileModuleContextInitialization() error {
 	// Note: if there's memory instruction in the function, memory instance must be non-nil.
 	// That is ensured by function validation at module instantiation phase, and that's
 	// why it is ok to skip the initialization if the module's memory instance is nil.
-	if c.ir.HasMemory {
+	if c.ir.HasMemory &&
+		// If this is the host function, our semantic is that host functions use the caller's memory.
+		// Therefore, we can skip the initialization step on memory here as at this point since it is
```
Comment (codefromthecrypt) [inline internal/engine/compiler/impl_arm64.go:4197]: same here, we aren't skipping anymore

File: internal/testing/enginetest/enginetest.go [hunk 3]
```diff
@@ -790,15 +790,22 @@ func setupCallMemTests(t *testing.T, e wasm.Engine, readMem *wasm.Code, fnlf exp
 
 	callReadMem := &wasm.Code{ // shows indirect calls still use the same memory
 		IsHostFunction: true,
-		Body:           []byte{wasm.OpcodeCall, 0, wasm.OpcodeEnd},
+		Body: []byte{
+			wasm.OpcodeCall, 1,
+			// On the return from the another host function,
+			// we should still be able to access the memory.
```
Comment (codefromthecrypt) [inline internal/testing/enginetest/enginetest.go:796]: cool!

**Review by codefromthecrypt — APPROVED** [review]
Thanks a lot!

**Pull request merged**